aspects/transaction_test.go

```diff
@@ -0,0 +1,328 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects_test
+
+import (
+	"errors"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/aspects"
+)
+
+type transactionTestSuite struct{}
+
+var _ = Suite(&transactionTestSuite{})
+
+type witnessReadWriter struct {
+	readCalled     int
+	writeCalled    int
+	bag            aspects.JSONDataBag
+	writtenDatabag aspects.JSONDataBag
+}
+
+func (w *witnessReadWriter) read() (aspects.JSONDataBag, error) {
+	w.readCalled++
+	return w.bag, nil
+}
+
+func (w *witnessReadWriter) write(bag aspects.JSONDataBag) error {
+	w.writeCalled++
+	w.writtenDatabag = bag
+	return nil
+}
+
+func (s *transactionTestSuite) TestSet(c *C) {
+	bag := aspects.NewJSONDataBag()
+	witness := &witnessReadWriter{bag: bag}
+	schema := aspects.NewJSONSchema()
+	tx, err := aspects.NewTransaction(witness.read, witness.write, schema)
+	c.Assert(err, IsNil)
+	c.Assert(witness.readCalled, Equals, 1)
+
+	err = tx.Set("foo", "bar")
+	c.Assert(err, IsNil)
+	c.Assert(witness.writeCalled, Equals, 0)
+
+	var value interface{}
+	err = witness.writtenDatabag.Get("foo", &value)
+	c.Assert(err, FitsTypeOf, &aspects.FieldNotFoundError{})
+}
+
+func (s *transactionTestSuite) TestCommit(c *C) {
+	witness := &witnessReadWriter{bag: aspects.NewJSONDataBag()}
+	schema := aspects.NewJSONSchema()
+	tx, err := aspects.NewTransaction(witness.read, witness.write, schema)
+	c.Assert(err, IsNil)
+	c.Assert(witness.readCalled, Equals, 1)
+
+	err = tx.Set("foo", "bar")
+	c.Assert(err, IsNil)
+	c.Assert(witness.readCalled, Equals, 1)
+	c.Assert(witness.writeCalled, Equals, 0)
+	c.Assert(witness.writtenDatabag, IsNil)
+
+	err = tx.Commit()
+	c.Assert(err, IsNil)
+
+	var value interface{}
+	err = witness.writtenDatabag.Get("foo", &value)
+	c.Assert(err, IsNil)
+
+	c.Assert(value, Equals, "bar")
+	c.Assert(witness.writeCalled, Equals, 1)
+
+	err = tx.Set("foo", "baz")
```

Comment:
To me the additional set and get as check seems unnecessary since we already established that set does call write, both in the set specific test and as part of the commit test.

Reply:
The goal is to make sure that subsequent writes aren't reflected in the written databag (since they're not committed). However, I pushed a new commit so transactions make less assumptions about databag isolation and the test covers this so I removed it here. Thanks